### PR TITLE
Update: Improve perf not traversing default ignored dirs (fixes #5679)

### LINKED ISSUE #5679: Extremely slow with antivirus
The company I work for requires that I run some horrible (but somehow popular) antivirus software on my PC which makes my `eslint .` take nearly 30 seconds to run on a very small project.

I did a little investigating and all of that slowness occurs in `glob-util.js` in the call to `glob.sync(pattern, globOptions)`. If I modify globOptions to include `ignore: 'node_modules/**'` my linting drops to one second.

It appears that crawling node_modules causes my antivirus to kick in for each file, and I'm assuming I can't be the only one with this problem. In any case, filtering out node_modules before the glob.sync call should certainly speed things up to some degree for everyone.

I understand that there is some complex logic involved with ignoring and unignoring paths so I see why all node_modules paths are included at this point in the code, but is there a chance this could be handled in some other way that would short-circuit the needless crawling? It would be a massive 99% speed increase, at least for me and anyone else in my position.

### PR DESCRIPTION
**Pull request by alberto**
Replaces #6215

### CODE REVIEW

**Comment by alberto** [comment]
Testing with PR on this repo:

```sh
$ time ./bin/eslint.js node_modules/

real	0m0.366s
user	0m0.329s
sys	0m0.050s
```

vs. master
```sh
$ time ./bin/eslint.js node_modules/

real	0m2.447s
user	0m2.199s
sys	0m0.276s
```

**Comment by gronke** [comment]
To entirely skip ignoring `bower_components` or `node_modules` when at least one pattern negates a path within them is a good compromise that will be an improvement for probably most of the use cases.

👍

**Comment by ilyavolodin** [comment]
LGTM

**Comment by samccone** [comment]
This is good, and is a good first pass improvement :) 👍

File: lib/ignored-paths.js [hunk 1]
```diff
@@ -188,4 +190,36 @@ IgnoredPaths.prototype.contains = function(filepath, category) {
 
 };
 
+/**
+ * Returns a list of dir patterns for glob to ignore
+ * @returns {Array} list of glob ignore patterns
+ */
+IgnoredPaths.prototype.ignoredFoldersGlobPatterns = function() {
```
Comment (nzakas) [inline lib/ignored-paths.js]: I'd call this `getIgnoredFoldersGlobPatterns` (our conventions have methods beginning with verbs)

File: lib/ignored-paths.js [hunk 2]
```diff
@@ -188,4 +190,36 @@ IgnoredPaths.prototype.contains = function(filepath, category) {
 
 };
 
+/**
+ * Returns a list of dir patterns for glob to ignore
+ * @returns {Array} list of glob ignore patterns
```
Comment (nzakas) [inline lib/ignored-paths.js]: Is this strings? If so, then use `{string[]}` for the type.

**Comment by nzakas** [comment]
Looks great, just a couple of very small stylistic things.

**Comment by alberto** [comment]
@nzakas PR updated

**Comment by nzakas** [comment]
LGTM. Awesome stuff!

**Comment by SimenB** [comment]
Any chance of doing a quick patch release of this?

Before: `node_modules/.bin/eslint .  10.76s user 2.53s system 76% cpu 17.312 total`
After: `node_modules/.bin/eslint .  0.91s user 0.11s system 81% cpu 1.250 total`

**Comment by ilyavolodin** [comment]
@SimenB We try to do releases every other Friday. So next release should be towards the end of this week.

**Pull request merged**